### PR TITLE
Expose MobileCode to python

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66304
* #66589
* __->__ #66171

This PR exposes MobileCode to python through pybind. THis is necessary when we land operator upgraders as a python file. We would have to run that file with MobileCode object to generate bytecode.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -13412,6 +13412,25 @@ def foo(a):
...
+    def test_mobile_code(self):
+        @torch.jit.script
+        def foo():
+            return torch.rand(2, 3)
+
+        mobileCode = torch._C.MobileCode(foo.graph, "foo")
+        instructions = mobileCode.instructions()
+        expected = [('LOADC', 1, 0),
+                    ('LOADC', 2, 0),
+                    ('LIST_CONSTRUCT', 0, 2),
```
Comment (tugsbayasgalan) [inline test/test_jit.py:13424]: @cccclai  do you think mobile will need to manually add offset to this number?

File: torch/csrc/jit/python/script_init.cpp [hunk 2]
```diff
@@ -778,6 +781,18 @@ void initJitScriptBindings(PyObject* module) {
   // NOLINTNEXTLINE(bugprone-unused-raii)
   py::class_<c10::Capsule>(m, "Capsule");
 
+  py::class_<MobileCode>(m, "MobileCode")
+      .def(py::init<std::shared_ptr<Graph>, std::string>())
+      .def("instructions", [](MobileCode& self) -> std::vector<IValue> {
```
Comment (iseeyuan) [inline torch/csrc/jit/python/script_init.cpp:786]: Is it just for instructions? As a complete frame would it also include other tables like constants, operators, types, etc?

**Comment by iseeyuan** [comment]
It's a nice feature to peek what the bytecode would look like.

There are several caveats to keep in mind: 
1. This feature produces the bytecode for current code. It does not reflect the bytecode saved in a model. 
2. Not sure if the upgrader bytecode is necessarily to be produced in python. Could it also be produced from C++?

**Comment by zhxchen17** [comment]
This implies we use python to compile a model to bytecodes, what's your "main compiler" going to look like?

File: torch/csrc/jit/python/script_init.cpp (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/python/script_init.cpp:786]: Yep I wasn't sure whats needed from mobile side. Those things can be included easily

**Comment by tugsbayasgalan** [comment]
> This implies we use python to compile a model to bytecodes, what's your "main compiler" going to look like?
Can you elaborate more on "main compiler"? @zhxchen17

**Comment by tugsbayasgalan** [comment]
> It's a nice feature to peek what the bytecode would look like.
> 
> There are several caveats to keep in mind:
> 
> 1. This feature produces the bytecode for current code. It does not reflect the bytecode saved in a model.
> 2. Not sure if the upgrader bytecode is necessarily to be produced in python. Could it also be produced from C++?

For 2) It was just done for convenience because you would need to just do something like `python upgraders.py` to generate the bytecodes

File: test/test_jit.py (hunk 1 above)
Comment (cccclai) [reply test/test_jit.py:13424]: Chatted offline. Let's also have type table. Maybe constant is needed, depending how bytecode looks like

**Comment by zhxchen17** [comment]
> > This implies we use python to compile a model to bytecodes, what's your "main compiler" going to look like?
> > Can you elaborate more on "main compiler"? @zhxchen17

@tugsbayasgalan I might miss something, what I mean is your upgrader definition file gonna look like the following or similar? (This is what I guess you want to do, but I'm not quite sure yet :) )
```
# upgrader_definitions.py
@torch.jit.script
def upgrader0(...):
    ...
@torch.jit.script
def upgrader1(...):
    ...

def export(f):
    mobile_code = MobileCode(f.graph())
    ...
    return (mobile_code, ...)

def main():
    defs = []
    for f in all_upgraders:
        def.append(export(f))
    ...
    return model(defs, ...) # output to somewhere
```
Then I guess instead of running `python upgrader_definitions.py` during pytorch build, we firstly build everything used to be, then use the pytorch we built to run this script, to get a separate module file, then "bundle" this module with the final mobile runtime?

File: test/test_jit.py (hunk 1 above)
Comment (tugsbayasgalan) [reply test/test_jit.py:13424]: sounds good.

**Comment by iseeyuan** [comment]
> > Then I guess instead of running `python upgrader_definitions.py` during pytorch build, we firstly build everything used to be, then use the pytorch we built to run this script, to get a separate module file, then "bundle" this module with the final mobile runtime?
> 
> Yeah this is what the file would look like. And your proposed idea makes lot of sense :). We can just finish building everything and then run this script to produce the yaml file. By doing so, mobile runtime would know that there is a generated yaml file ready. @gmagogsfm @iseeyuan @cccclai what do you guys think?

Putting upgraders in a module sounds a straightforward way to implement. Eventually what mobile side needs is a static table. The keys are the old op names and the values are the mobile::functions/frames. We could start from loading this bytecode module and populating the table. The caveat is that the deserialization/loading may take some overhead on cold start. It would be nice (in the future) to move part of it to build time. @cccclai How do you think?

File: test/test_jit.py (hunk 1 above)
Comment (cccclai) [reply test/test_jit.py:13424]: Sharing an example bytecode table, we can probably use the same format
```
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('LOAD', 2, 0),
     ('LOADC', 2, 0),
     ('LOADC', 1, 0),
     ('LIST_CONSTRUCT', 0, 1),
     ('OP', 0, 0),
     ('LOADC', 1, 0),
     ('LIST_CONSTRUCT', 1, 1),
     ('OP', 1, 0),
     ('TUPLE_CONSTRUCT', 2, 0),
     ('DICT_CONSTRUCT', 2, 2),
     ('LOAD', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 2, 0),
     ('OP', 3, 0),
     ('LOADC', 3, 0),
     ('OP', 4, 0),
     ('LIST_CONSTRUCT', 3, 1),
     ('OP', 5, 0),
     ('LOAD', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 6, 0),
     ('OP', 7, 0),
     ('LOADC', 3, 0),
     ('OP', 8, 0),
     ('LIST_CONSTRUCT', 4, 1),
     ('OP', 9, 0),
     ('TUPLE_CONSTRUCT', 2, 0),
     ('LOAD', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 10, 0),
     ('OP', 11, 0),
     ('LOADC', 3, 0),
     ('OP', 12, 0),
     ('LIST_CONSTRUCT', 5, 1),
     ('OP', 13, 0),
     ('LOAD', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 14, 0),
     ('OP', 15, 0),
     ('LOADC', 3, 0),
     ('OP', 16, 0),
     ('LIST_CONSTRUCT', 6, 1),
     ('OP', 17, 0),
     ('TUPLE_CONSTRUCT', 2, 0),
     ('LOAD', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 18, 0),
     ('OP', 19, 0),
     ('LOADC', 3, 0),
     ('OP', 20, 0),
     ('LIST_CONSTRUCT', 7, 1),
     ('OP', 21, 0),
     ('MOVE', 2, 0),
     ('LOADC', 3, 0),
     ('OP', 22, 0),
     ('OP', 23, 0),
     ('LOADC', 3, 0),
     ('OP', 24, 0),
     ('LIST_CONSTRUCT', 8, 1),
     ('OP', 25, 0),
     ('TUPLE_CONSTRUCT', 2, 0),
     ('LOADC', 2, 0),
     ('LOADC', 1, 0),
     ('LIST_CONSTRUCT', 9, 1),
     ('OP', 26, 0),
     ('LOADC', 1, 0),
     ('LIST_CONSTRUCT', 10, 1),
     ('OP', 27, 0),
     ('LOADC', 1, 0),
     ('LIST_CONSTRUCT', 11, 1),
     ('OP', 28, 0),
     ('TUPLE_CONSTRUCT', 3, 0),
     ('DICT_CONSTRUCT', 12, 2),
     ('NAMED_TUPLE_CONSTRUCT', 13, 6),
     ('RET', 0, 0))),
   ('operators',
    (('aten::zeros', '', 1),
     ('aten::zeros', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::ones', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::ones', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::ones', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::ones', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::zeros', '', 1),
     ('prim::TupleIndex', '', 2),
     ('aten::size', '', 1),
     ('aten::__getitem__', 't', 2),
     ('aten::ones', '', 1),
     ('aten::zeros', '', 1),
     ('aten::zeros', '', 1),
     ('aten::zeros', '', 1))),
   ('constants', (None, 1, -1, 0)),
   ('types',
    ('List[int]',
     'List[int]',
     'Dict[int, Tuple[Tensor, Tensor]]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'List[int]',
     'Dict[int, Tuple[Tensor, Tensor, Tensor]]',)),
   ('register_size', 2)),
```

**Comment by tugsbayasgalan** [comment]
> > > Then I guess instead of running `python upgrader_definitions.py` during pytorch build, we firstly build everything used to be, then use the pytorch we built to run this script, to get a separate module file, then "bundle" this module with the final mobile runtime?
> > 
> > 
> > Yeah this is what the file would look like. And your proposed idea makes lot of sense :). We can just finish building everything and then run this script to produce the yaml file. By doing so, mobile runtime would know that there is a generated yaml file ready. @gmagogsfm @iseeyuan @cccclai what do you guys think?
> 
> Putting upgraders in a module sounds a straightforward way to implement. Eventually what mobile side needs is a static table. The keys are the old op names and the values are the mobile::functions/frames. We could start from loading this bytecode module and populating the table. The caveat is that the deserialization/loading may take some overhead on cold start. It would be nice (in the future) to move part of it to build time. @cccclai How do you think?

I think it doesn't need to be a module. We can just generate the yaml file from `python upgrader_definitions.py` that can be accessible from mobile.

**Comment by cccclai** [comment]
> > > Then I guess instead of running `python upgrader_definitions.py` during pytorch build, we firstly build everything used to be, then use the pytorch we built to run this script, to get a separate module file, then "bundle" this module with the final mobile runtime?
> > 
> > 
> > Yeah this is what the file would look like. And your proposed idea makes lot of sense :). We can just finish building everything and then run this script to produce the yaml file. By doing so, mobile runtime would know that there is a generated yaml file ready. @gmagogsfm @iseeyuan @cccclai what do you guys think?
> 
> Putting upgraders in a module sounds a straightforward way to implement. Eventually what mobile side needs is a static table. The keys are the old op names and the values are the mobile::functions/frames. We could start from loading this bytecode module and populating the table. The caveat is that the deserialization/loading may take some overhead on cold start. It would be nice (in the future) to move part of it to build time. @cccclai How do you think?

Very good point!

File: test/test_jit.py (hunk 1 above)
Comment (tugsbayasgalan) [reply test/test_jit.py:13424]: ok

**Pull request closed without merging**